feat: support maxmemory-clients config

AntiTopQuark opened the pull request:
relate to https://github.com/apache/kvrocks/issues/2284
Implemented a new maxmemory-clients configuration to enable client eviction based on overall memory usage, preventing potential OOM issues due to excessive client connections.

AntiTopQuark committed:
feat: support maxmemory-clients

AntiTopQuark committed:
feat: support maxmemory-clients

c8ef committed:
ci: add Debian environment to CI workflow (#2510)

AntiTopQuark committed:
update

PragmaTwice commented on src/server/server.cc:
```diff
@@ -1136,12 +1141,13 @@ void Server::GetStatsInfo(std::string *info) {
   string_stream << "keyspace_misses:" << db_stats->keyspace_misses << "\r\n";
 
-  {
+  {https://www.anyknew.com/go/10112884
```
hmm

AntiTopQuark replied:
I'm sorry for the mistake

AntiTopQuark replied:
I'll review the code again shortly.

AntiTopQuark committed:
update

PokIsemaine committed:
feat(config): add `txn_context_enabled` to allow to enable the transaction feature (#2506)

Co-authored-by: mwish <maplewish117@gmail.com>

PokIsemaine committed:
fix(test): remove meaningless `list-max-ziplist-size` (#2517)

git-hulk committed:
fix(replication): potential deadlock  when switching master frequently (#2516)

This closes #2512.

Currently, the replication thread will wait for the worker's exclusive guard stop before closing db.
But it now stops the worker from running new commands after acquiring the worker's exclusive guard,
and it might cause deadlock if switches at the same time.

The following steps will show how it may happen:

- T0: client A sent `slaveof MASTER_IP0 MASTER_PORT0`, then the replication thread was started and waiting for the exclusive guard.

- T1: client B sent `slaveof MASTER_IP1 MASTER_PORT1` and `AddMaster` will stop the previous replication thread, which is waiting for the exclusive guard. But the exclusive guard is acquired by the current thread.

The workaround is also straightforward, just stop workers from running new commands by enabling `is_loading_` to 
true before acquiring the lock in the replication thread.

AntiTopQuark committed:
update

AntiTopQuark committed:
fix(typo): `filed -> field` (#2519)

AntiTopQuark committed:
update

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients

AntiTopQuark committed:
fix data race error

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients

AntiTopQuark committed:
update unit case

caipengbo commented:
I want to discuss how we can implement this feature. My idea is to count client memory within each worker (using `libevent`'s callback to count `evbuffer`) to avoid locking across workers.  @AntiTopQuark @PragmaTwice @git-hulk @mapleFU @torwig

git-hulk commented:
> I want to discuss how we can implement this feature. My idea is to count client memory within each worker (using `libevent`'s callback to count `evbuffer`) to avoid locking across workers. @AntiTopQuark @PragmaTwice @git-hulk @mapleFU @torwig

Yes, as mentioned in [previous comment](https://github.com/apache/kvrocks/pull/2498#issuecomment-2311996885). It would be better to limit the connection output first by checking the libevent buffer. And for the `maxmemory-clients`, I think we don't need to count every byte, just checking the libevent input and output buffer is good to me as well.

AntiTopQuark commented:
> > I want to discuss how we can implement this feature. My idea is to count client memory within each worker (using `libevent`'s callback to count `evbuffer`) to avoid locking across workers. @AntiTopQuark @PragmaTwice @git-hulk @mapleFU @torwig
> 
> Yes, as mentioned in [previous comment](https://github.com/apache/kvrocks/pull/2498#issuecomment-2311996885). It would be better to limit the connection output first by checking the libevent buffer. And for the `maxmemory-clients`, I think we don't need to count every byte, just checking the libevent input and output buffer is good to me as well.

I have modified the code and moved the `evictionClients` function into `Worker`.
On another note, I believe that `maxmemory-clients` is used to limit the total size of connections. If you want to limit the size of the output buffer, you can use the client-output-buffer-limit configuration option. I will submit a PR for this in the future.

AntiTopQuark committed:
update

AntiTopQuark committed:
update

mapleFU commented:
>  want to discuss how we can implement this feature. My idea is to count client memory within each worker (using libevent's callback to count evbuffer) to avoid locking across workers. 

Yeah I think holding a lock is harmful here, I can accept some untimely memory change, but global checking is harmful...

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients

AntiTopQuark committed:
update

AntiTopQuark committed:
update

AntiTopQuark committed:
u

AntiTopQuark committed:
update code style

AntiTopQuark committed:
update code style

PragmaTwice commented on src/server/redis_connection.cc:
```diff
@@ -559,4 +562,37 @@ void Connection::ResetMultiExec() {
 }
 
+size_t Connection::GetConnectionMemoryUsed() const {
+  size_t total_memory = sizeof(*this);  // 包含所有成员变量的静态内存大小
```
please dont include chinese comments.

AntiTopQuark replied:
done

PragmaTwice commented on src/server/redis_connection.cc:
```diff
@@ -559,4 +562,37 @@ void Connection::ResetMultiExec() {
+size_t Connection::GetConnectionMemoryUsed() const {
+  size_t total_memory = sizeof(*this);  // 包含所有成员变量的静态内存大小
+
+  total_memory += name_.capacity();
+  total_memory += ns_.capacity();
+  total_memory += ip_.capacity();
+  total_memory += announce_ip_.capacity();
+  total_memory += addr_.capacity();
+  total_memory += last_cmd_.capacity();
+  total_memory += output_buffer_.capacity();
+  total_memory += slave_output_buffer_.capacity();
+  total_memory += evbuffer_get_length(Output()) + evbuffer_get_length(Input());
+
+  for (const auto &channel : subscribe_channels_) {
+    total_memory += channel.capacity();
+  }
+  for (const auto &pattern : subscribe_patterns_) {
+    total_memory += pattern.capacity();
+  }
+  for (const auto &channel : subscribe_shard_channels_) {
+    total_memory += channel.capacity();
+  }
+  for (const auto &cmd : multi_cmds_) {
+    total_memory += cmd.capacity();
+  }
+
+  if (saved_current_command_) {
+    total_memory += saved_current_command_->GetMemoryUsage();
+  }
```
The code seems hard to maintain and also not like a precise estimation of memory usage.

I'm wondering if it's really necessary to limit the connection memory usage..

PragmaTwice commented:
I believe implementing an elegant solution to track and restrict client connection memory usage is challenging and could reduce code maintainability. 

Instead, I suggest exploring alternative approaches, like limiting the number of clients.

AntiTopQuark committed:
update code style

AntiTopQuark commented:
> I believe implementing an elegant solution to track and restrict client connection memory usage is challenging and could reduce code maintainability.
> 
> Instead, I suggest exploring alternative approaches, like limiting the number of clients.



> I believe implementing an elegant solution to track and restrict client connection memory usage is challenging and could reduce code maintainability.
> 
> Instead, I suggest exploring alternative approaches, like limiting the number of clients.

When I looked at Redis previously, it uses the maxclients configuration option to limit the number of clients, and maxmemory-clients to limit the total memory used by connections. If you only want to limit the output buffer, you can use client-output-buffer-limit to restrict the usage per connection.

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients

AntiTopQuark commented:
@PragmaTwice  @git-hulk  Hi, should we continue pushing this PR forward, or can it be closed?

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients

AntiTopQuark committed:
Merge branch 'unstable' into limit_maxmemory_clients
